tensorboard/program.py

```diff
@@ -386,13 +394,17 @@ def _fix_mime_types(self):
         mimetypes.add_type("text/html", ".html")
 
     def _make_data_ingester(self):
+        """Determines the right data ingester, starts it, and returns it."""
         flags = self.flags
         if flags.grpc_data_provider:
-            return server_ingester.ExistingServerDataIngester(
+            ingester = server_ingester.ExistingServerDataIngester(
                 flags.grpc_data_provider,
                 channel_creds_type=flags.grpc_creds_type,
             )
-        if _should_use_data_server(flags.load_fast, flags.logdir):
+            ingester.start()
+            return ingester
+
+        if flags.load_fast in ("auto", "true"):
```

Comment:
Optional, but I find the control flow here to have gotten pretty convoluted.  What about just putting `server_ingester.get_server_binary()`, `ingester = server_ingester.SubprocessServerDataIngester()`, and `ingester.start()` all into a single common helper (say `start_data_server_and_ingester()`) and then splitting out the two cases for "auto" and "true"?  It does mean you have a broader try-except block but since the exceptions are typed that doesn't seem like the end of the world, and it would simplify the logic to:

```python
if flags.load_fast == "true":
    try:
        return start_data_server_and_ingester()
    except server_ingester.NoDataServerError as e:
        sys.stderr.write("Option --load_fast=true not available" ... )
        sys.exit(1)
    except server_ingester.DataServerStartupError as e:
        sys.stderr.write(_DATA_SERVER_STARTUP_ERROR_MESSAGE_TEMPLATE % e)
        sys.exit(1)

if flags.load_fast == "auto":
    try:
        return start_data_server_and_ingester()
    except (server_ingester.NoDataServerError, server_ingester.DataServerStartupError) as e:
        logger.info("Data server error: %s; falling back to multiplexer", e)

ingester = local_ingester.LocalDataIngester(flags)
ingester.start()
return ingester
```

Reply:
I like it; thanks. Done.